include/sharg/parser.hpp

```diff
@@ -182,27 +181,30 @@ class parser
      * \stableapi{Since version 1.0.}
      */
     parser(std::string const & app_name,
-           int const argc,
-           char const * const * const argv,
+           std::vector<std::string> const & arguments,
            update_notifications version_updates = update_notifications::on,
            std::vector<std::string> subcommands = {}) :
         version_check_dev_decision{version_updates},
-        subcommands{std::move(subcommands)}
+        original_arguments{arguments}
     {
-        for (auto & sub : this->subcommands)
-        {
-            if (!std::regex_match(sub, app_name_regex))
-            {
-                throw design_error{"The subcommand name must only contain alpha-numeric characters or '_' and '-' "
-                                   "(regex: \"^[a-zA-Z0-9_-]+$\")."};
-            }
-        }
-
         info.app_name = app_name;
 
-        init(argc, argv);
+        for (auto const & subcommand : subcommands)
+            std::ignore /* function is nodisard */ = add_subcommand(subcommand);
```

Comment:
This doesn't work for this approach, ctor would need to be removed.
Or we need to store whether the ctor did have subcommands:
* If no, we allow calling `add_subcommand`,
* otherwise we throw when calling `add_subcommand`

If any of the subcommands in the ctor are present, `sub_parser` is set to a value `!= nullptr`. Then any `add_subcommand` would return that pointer.